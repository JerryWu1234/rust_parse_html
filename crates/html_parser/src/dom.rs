use std::collections::HashMap;

#[derive(Debug)]
pub struct Node {
  pub node_type: NodeType,
}

#[derive(Debug)]
pub struct ElementPreproty {
  pub tag_name: String,
  pub attributes: HashMap<String, String>,
  pub children: Vec<Node>,
}

#[derive(Debug)]
pub enum NodeType {
  Text(String),
  Element(ElementPreproty),
}

pub fn create_element(
  tag_name: String,
  attributes: HashMap<String, String>,
  children: Vec<Node>,
) -> Node {
  Node {
    node_type: (NodeType::Element(ElementPreproty {
      tag_name,
      attributes,
      children,
    })),
  }
}

pub fn create_text(text: String) -> Node {
  Node {
    node_type: (NodeType::Text(text)),
  }
}
