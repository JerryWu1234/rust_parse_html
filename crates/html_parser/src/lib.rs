pub mod dom;
// use std::{collections::HashMap, hash::Hash};

use std::{collections::HashMap, hash::Hash};

pub use dom::*;

#[derive(Debug)]
pub struct Parser {
  pub pos: usize,
  pub input: String,
}

impl Parser {
  pub fn parse_nodes(&mut self) -> Vec<Node> {
    let mut node: Vec<Node> = vec![];
    loop {
      self.consume_whitespace();
      println!(">>{:?}", self.input[self.pos..].to_string());
      if self.eof() || self.input[self.pos..].starts_with("</") {
        break;
      }
      node.push(self.parse_node())
    }
    node
  }

  pub fn parse_node(&mut self) -> Node {
    match self.next_char() {
      '<' => self.cusume_element(),
      _ => self.cusume_text(),
    }
  }

  pub fn cusume_text(&mut self) -> Node {
    dom::create_text(self.iteration(|c| c != '>'))
  }

  pub fn cusume_element(&mut self) -> Node {
    assert_eq!('<', self.consume_char());
    let tag_name = self.tag_name();
    self.consume_whitespace();
    let attributes = self.consume_attributes();
    let children = self.parse_nodes();
    dom::create_element(tag_name, attributes, children)
  }

  pub fn consume_attributes(&mut self) -> HashMap<String, String> {
    let mut attrs = HashMap::new();
    loop {
      self.consume_whitespace();
      if self.consume_char() == '>' {
        break;
      }
      let (key, value) = self.get_attributes();
      attrs.insert(key, value);
    }
    attrs
  }

  pub fn get_attributes(&mut self) -> (String, String) {
    let key = self.iteration(getStrings);
    assert_eq!('=', self.consume_char());
    let open_quote = self.consume_char();
    assert!(open_quote == '"' || open_quote == '\'');
    let value = self.iteration(|c| c != open_quote);
    assert_eq!('"', self.consume_char());

    (key, value)
  }

  pub fn tag_name(&mut self) -> String {
    self.iteration(getStrings)
  }

  pub fn consume_whitespace(&mut self) {
    self.iteration(char::is_whitespace);
  }

  pub fn next_char(&mut self) -> char {
    self.input[self.pos..].chars().next().unwrap()
  }

  pub fn iteration<F>(&mut self, callback: F) -> String
  where
    F: Fn(char) -> bool,
  {
    let mut value = String::from("");
    while !self.eof() && callback(self.next_char()) {
      value.push(self.consume_char())
    }
    value
  }

  pub fn consume_char(&mut self) -> char {
    let (index, char) = self.input[self.pos..].char_indices().next().unwrap();
    self.pos += index + 1;
    println!("{},>>>打印当前循环的char", char);
    char
  }

  pub fn eof(&self) -> bool {
    self.pos >= self.input.len()
  }
}

fn getStrings(v: char) -> bool {
  match v {
    'a'..='z' | 'A'..='Z' | '0'..='9' => true,
    _ => false,
  }
}
