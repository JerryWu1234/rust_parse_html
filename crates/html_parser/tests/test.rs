use html_parser::dom;
use html_parser::Parser;
use insta::assert_debug_snapshot;

#[test]
fn it_works() {
    let m = dom::create_text("example".to_string());
    assert_debug_snapshot!(m);
}

#[test]
fn it_parser() {
    let c = Parser {
        pos: 1,
        input: "<div></div>".to_string(),
    }
    .next_char();
    assert_eq!('d', c)
}

#[test]
fn testiterator() {
    let mut c = Parser {
        pos: 0,
        input: "   <div></div>".to_string(),
    };

    c.iteration(|c| match c {
        'a'..='z' | 'A'..='Z' | '0'..='9' => false,
        _ => true,
    });
    let v = c.consume_char();
    assert_eq!('d', v)
}

#[test]
fn test_attribute() {
    let mut c = Parser {
        pos: 0,
        input: "<div id=\"myid\">
        <p>3223</p>
        </div>".to_string(),
    };
    let m = c.parse_nodes();
    assert_debug_snapshot!(m);
}
