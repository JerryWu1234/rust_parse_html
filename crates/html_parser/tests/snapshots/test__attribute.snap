---
source: crates/html_parser/tests/test.rs
assertion_line: 45
expression: m

---
[
    Node {
        node_type: Element(
            ElementPreproty {
                tag_name: "div",
                attributes: {
                    "d": "myid",
                },
                children: [],
            },
        ),
    },
]
